src/samplers/mh.jl

```diff
@@ -0,0 +1,243 @@
+  space     ::  Set       # sampling space, emtpy means all
+  gid       ::  Int       # group ID
+  function MH(n_iters::Int, space...)
+    new_space = Set()
+    proposals = Dict{Symbol,Any}()
+
+    # parse random variables with their hypothetical proposal
+    for element in space
+        if isa(element, Symbol)
+          push!(new_space, element)
+        else
+          @assert isa(element[1], Symbol) "[MH] ($element[1]) should be a Symbol. For proposal, use the syntax MH(N, (:m, (x) -> Normal(x, 0.1)))"
+          push!(new_space, element[1])
+          proposals[element[1]] = element[2]
+        end
+    end
+
+    new(n_iters, proposals, Set(new_space), 0)
+  end
+  MH(alg::MH, new_gid::Int) = new(alg.n_iters, alg.proposals, alg.space, new_gid)
+end
+
+Sampler(alg::MH) = begin
+  alg_str = "MH"
+
+  # Sanity check for space
+  if alg.gid == 0 && !isempty(alg.space)
+    @assert issubset(Turing._compiler_[:pvars], alg.space) "[$alg_str] symbols specified to samplers ($alg.space) doesn't cover the model parameters ($(Turing._compiler_[:pvars]))"
+    if Turing._compiler_[:pvars] != alg.space
+      warn("[$alg_str] extra parameters specified by samplers don't exist in model: $(setdiff(alg.space, Turing._compiler_[:pvars]))")
+    end
+  end
+
+  info = Dict{Symbol, Any}()
+  info[:accept_his] = []
+  info[:total_eval_num] = 0
+  info[:proposal_ratio] = 0.0
+  info[:prior_prob] = 0.0
+  info[:violating_support] = false
+
+  Sampler(alg, info)
+end
+
+propose(model::Function, spl::Sampler{MH}, vi::VarInfo) = begin
+  spl.info[:proposal_ratio] = 0.0
+  spl.info[:prior_prob] = 0.0
+  spl.info[:violating_support] = false
+  runmodel(model, vi ,spl)
+end
+
+step(model::Function, spl::Sampler{MH}, vi::VarInfo, is_first::Bool) = begin
+  if is_first
+    push!(spl.info[:accept_his], true)
+    vi
+
+  else
+    if spl.alg.gid != 0 # Recompute joint in logp
+      runmodel(model, vi, nothing)
+    end
+    old_θ = copy(vi[spl])
+    old_logp = getlogp(vi)
+
+    dprintln(2, "Propose new parameters from proposals...")
+    propose(model, spl, vi)
+
+    dprintln(2, "computing accept rate α...")
+    α = getlogp(vi) - old_logp + spl.info[:proposal_ratio]
+
+    dprintln(2, "decide wether to accept...")
+    if log(rand()) < α && !spl.info[:violating_support]  # accepted
+      push!(spl.info[:accept_his], true)
+    else                      # rejected
+      push!(spl.info[:accept_his], false)
+      vi[spl] = old_θ         # reset Θ
+      setlogp!(vi, old_logp)  # reset logp
+    end
+
+    vi
+  end
+end
+
+function sample(model::Function, alg::MH;
+                save_state=false,         # flag for state saving
+                resume_from=nothing,      # chain to continue
+                reuse_spl_n=0,            # flag for spl re-using
+                )
+
+  spl = reuse_spl_n > 0 ?
+        resume_from.info[:spl] :
+        Sampler(alg)
+  alg_str = "MH"
+
+  # Initialization
+  time_total = zero(Float64)
```

Comment:
The main purpose of using `zero(T)` is to avoid type conversion and promotion (which cost time). It seems that `zero()` makes no difference from `0.0` here () BUT it's OK to keep it.